### ISSUE TITLE
thread-instruction-count: tidelta / ticount don't exist in tracing file

### ISSUE THREAD

**Issue by mbalabash** [puppeteer/puppeteer#5223]
**Environment 1**
puppeteer-core: 2.0.0
docker image: node:12-slim
os version: macOS 10.14.6
node.js version: 12
chrome version: 77/78

**Environment 2**
puppeteer-core: 2.0.0
virtualisation: virtualbox 6
os version: Ubuntu 18.04
node.js version: 12
chrome version: 77

**What steps will reproduce the problem?**
According to [this message](https://github.com/GoogleChrome/puppeteer/pull/4591#issuecomment-502340394). Puppeteer provides [RTI](https://chromium-review.googlesource.com/c/chromium/src/+/1589672) data in trace file when flags: `--enable-thread-instruction-count` and `--no-sandbox` are used. Information should be placed in `tidelta` and `ticount` properties.

Case:
```js
const puppeteer = require('puppeteer')

;(async () => {
  const browser = await puppeteer.launch({
    args: ['--no-sandbox', '--enable-thread-instruction-count'],
  })
  const page = await browser.newPage()

  await page.tracing.start({ path: './trace.json' })
  await page.goto('https://www.example.com/')
  await page.tracing.stop()

  await browser.close()
})()
```
[Dockerfile](https://github.com/mbalabash/estimo/blob/rti/Dockerfile)
[Chrome-Detect&Downloader](https://github.com/mbalabash/estimo/blob/rti/scripts/chromeDetection.js)

**What is the expected result?**
`tidelta` exist in trace.json and parameter not empty.

**What happens instead?**
`tidelta` doesn't exist in trace.json.

**What a goal to use this feature?**
I want to use RTI data for [Web Perf Testing](https://twitter.com/aslushnikov/status/1121816765518897152) in tools like [size-limit](https://github.com/ai/size-limit), [estimo](https://github.com/mbalabash/estimo), etc.

**PS**
I tried this case with docker image node:10-slim, chrome 77 and chrome 78.
But `tidelta` always didn't exist in trace file.

**Comment by mbalabash** [comment]
I got more information about this problem.
After launching [script](https://github.com/mbalabash/estimo/blob/rti/src/create-chrome-trace.js#L76) with `dumpio: true` I got chromium log: 
```
DevTools listening on ws://127.0.0.1:39853/devtools/browser/f1ed78b3-cd2b-418c-a7bc-79a24a7021d3
[0109/131032.904194:WARNING:ipc_message_attachment_set.cc(49)] MessageAttachmentSet destroyed with unconsumed attachments: 0/1
[0109/131032.904284:ERROR:command_buffer_proxy_impl.cc(124)] ContextResult::kTransientFailure: Failed to send GpuChannelMsg_CreateCommandBuffer.
[0109/131032.918871:WARNING:ipc_message_attachment_set.cc(49)] MessageAttachmentSet destroyed with unconsumed attachments: 0/1
[0109/131032.918890:ERROR:command_buffer_proxy_impl.cc(124)] ContextResult::kTransientFailure: Failed to send GpuChannelMsg_CreateCommandBuffer.
tracing_service_impl.cc: Configured tracing, #sources:2, duration:0 ms, #buffers:1, total buffer size:102400 KB, total sessions:1
[0109/131032.951528:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.951529:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.951764:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.952957:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.954586:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.954995:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.955406:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.957047:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960385:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960369:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960750:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960873:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960894:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.960788:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131032.961589:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131033.252945:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131033.285899:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131033.333724:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[0109/131033.624007:WARNING:cors_url_loader_factory.cc(214)] |cors_exempt_headers| contains unexpected key: Purpose
[0109/131033.806456:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
tracing_service_impl.cc: Tracing session 1 ended, total sessions:0
```

**perf_event_open failed, omitting instruction counters** - seems like by default in Docker environment new API doesn't have access to required [syscall - perf_event_open](https://cs.chromium.org/chromium/src/base/trace_event/thread_instruction_count.cc?sq=package:chromium&g=0&l=62).

**Without this fix, we cannot use new awesome API on CI.**
Any thoughts about how we can fix it?

**Comment by mbalabash** [comment]
According to [this article](https://blog.alicegoldfuss.com/enabling-perf-in-kubernetes/) and Docker docs ([link1](https://docs.docker.com/engine/security/seccomp/#significant-syscalls-blocked-by-the-default-profile), [link2](https://docs.docker.com/engine/reference/run/#additional-groups)) there are three ways to allow this syscall:

- Add special linux capability (`--cap-add SYS_ADMIN`) 
- Run Docker image in privileged mode (`--privileged`)
- Extend [default Docker seccomp](https://github.com/moby/moby/blob/master/profiles/seccomp/default.json) profile, add `perf_event_open` to `SCMP_ACT_ALLOW` group, remove `perf_event_open` from group with `"caps": ["CAP_SYS_ADMIN"]` and  use it as custom seccomp profile.

**Comment by mbalabash** [comment]
I have tested all these ways in a few different environments:
- macOS / LinuxKit VM / Docker / node:12-slim
- Win10 / VirtualBox / ubuntu 18.04 lts
- VMware / CentOS

But neither didn't work for me (still same chromium error output).

**Comment by mbalabash** [comment]
One gets the impression that this API can only work from under bare Linux, and any virtualization breaks it.

Can you help me to find out about this problem?

**Comment by mbalabash** [comment]
Hello everyone!
I want to ask your help guys.

If you are using Linux distro without any virtualization, you can help with testing this issue.
(Currently, I can't do it by myself, unfortunately)

**What you can do**:

1. `git clone -b rti https://github.com/mbalabash/estimo`

2. `yarn install` (_It might take a while_)

3. `yarn test:launch`

4. Copy+Paste console output from the latest command and include it in your comment

5. Copy content from json file in `path/to/estimo/temp/some-generic-id.json`

   5.1. Create gist and paste content from previous step

   5.2. Provide link to this gist in your comment

6. Provide your Linux distro name and version

7. Post your comment and @ tag me there

Thanks to everyone who will help.

**Comment by ai** [comment]
```
estimo on  rti [!] 
➜ yarn test:launch
yarn run v1.21.1
$ ESTIMO_DEBUG=true node ./test/__mock__/test-launch.js
/home/ai/estimo/test/__mock__

[estimo]: Found next js files: /home/ai/estimo/test/__mock__/19kb.js


[estimo]: Found next web pages: 


[js-mode]: ------------------------------------------
[js-mode]: Preparing file: /home/ai/estimo/test/__mock__/19kb.js

[estimo]: The file has been written: /home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html

[js-mode]: Creating html content for js file: /home/ai/estimo/test/__mock__/19kb.js
[js-mode]: Js file name: 19kb.js
[js-mode]: Html file: /home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html
[js-mode]: Url to html file: file:///home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html
[js-mode]: Html content: 
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Estimo Template</title>
  </head>
  <body>
    <script src="/home/ai/estimo/test/__mock__/19kb.js"></script>
    <h1>Estimo</h1>
  </body>
</html>
[js-mode]: ------------------------------------------


DevTools listening on ws://127.0.0.1:34275/devtools/browser/3867a6ed-b816-4b38-96e8-821a50e98505
tracing_service_impl.cc: Configured tracing, #sources:2, duration:0 ms, #buffers:1, total buffer size:102400 KB, total sessions:1
tracing_service_impl.cc: Tracing session 1 ended, total sessions:0

[js-mode]: Next javascript resources has been prepared: [{"name":"19kb.js","url":"file:///home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html","html":"/home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html","trace":"/home/ai/estimo/temp/ZcNLFX8-IUKh1r_fnbkAd.json"}]


[report]: Processing resource: {"name":"19kb.js","url":"file:///home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html","html":"/home/ai/estimo/temp/sPzYoRgL34w3exHEdtDio.html","trace":"/home/ai/estimo/temp/ZcNLFX8-IUKh1r_fnbkAd.json"}


[js-mode]: Got reports for js files: [{"name":"19kb.js","parseHTML":1.36,"styleLayout":4.55,"paintCompositeRender":0.47,"scriptParseCompile":2.14,"scriptEvaluation":9.82,"javaScript":11.96,"garbageCollection":0,"other":6.63,"total":24.97}]


[estimo]: Result reports: [{"name":"19kb.js","parseHTML":1.36,"styleLayout":4.55,"paintCompositeRender":0.47,"scriptParseCompile":2.14,"scriptEvaluation":9.82,"javaScript":11.96,"garbageCollection":0,"other":6.63,"total":24.97}]

[
  {
    name: '19kb.js',
    parseHTML: 1.36,
    styleLayout: 4.55,
    paintCompositeRender: 0.47,
    scriptParseCompile: 2.14,
    scriptEvaluation: 9.82,
    javaScript: 11.96,
    garbageCollection: 0,
    other: 6.63,
    total: 24.97
  }
]
Done in 0.54s
```

JSON: https://gist.github.com/ai/a5d31776e407ac701efd5c7670e0eeda

Fedora 31

@mbalabash

**Comment by mbalabash** [comment]
Ok. It definitely works on a bare Linux machine.

So, I have two ideas:

- I mess up with granting permissions to docker (for executing perf_event_open)

- Something else wrong behind perf_event_open

**Comment by mbalabash** [comment]
I tried to run `perf` tool in docker container (in --privileged mode) manually and got this output:

```sh
# perf stat -d ls
CHANGELOG.md  Dockerfile  LICENSE  README.md  chrome.json  index.js  node_modules  package.json  scripts  seccomp-conf.json  src  temp  test  yarn.lock

 Performance counter stats for 'ls':

          1.371443      task-clock (msec)         #    0.561 CPUs utilized          
                 0      context-switches          #    0.000 K/sec                  
                 0      cpu-migrations            #    0.000 K/sec                  
                79      page-faults               #    0.058 M/sec                  
   <not supported>      cycles                                                      
   <not supported>      instructions                                                
   <not supported>      branches                                                    
   <not supported>      branch-misses                                               
   <not supported>      L1-dcache-loads                                             
   <not supported>      L1-dcache-load-misses                                       
   <not supported>      LLC-loads                                                   
   <not supported>      LLC-load-misses                                             

       0.002445000 seconds time elapsed
```

Seems like linuxkit/docker doesn't support some of the processor's hardware performance counters, which `perf` uses.

I also tested all ways which I used to grant permissions to docker and all of them work well. So, this issue not about permissions for executing perf_event_open anymore.

**Comment by mbalabash** [comment]
My `perf list cpu` output:

```sh
# perf list cpu

List of pre-defined events (to be used in -e):

  cpu-clock                                          [Software event]
  cpu-migrations OR migrations                       [Software event]

  cpuhp:cpuhp_enter                                  [Tracepoint event]
  cpuhp:cpuhp_exit                                   [Tracepoint event]
  cpuhp:cpuhp_multi_enter                            [Tracepoint event]
  kmem:mm_page_pcpu_drain                            [Tracepoint event]
  power:cpu_frequency                                [Tracepoint event]
  power:cpu_idle                                     [Tracepoint event]
  syscalls:sys_enter_getcpu                          [Tracepoint event]
  syscalls:sys_exit_getcpu                           [Tracepoint event]
  thermal:thermal_power_cpu_get_power                [Tracepoint event]
  thermal:thermal_power_cpu_limit                    [Tracepoint event]
  xen:xen_cpu_load_idt                               [Tracepoint event]
  xen:xen_cpu_set_ldt                                [Tracepoint event]
  xen:xen_cpu_write_gdt_entry                        [Tracepoint event]
  xen:xen_cpu_write_idt_entry                        [Tracepoint event]
  xen:xen_cpu_write_ldt_entry                        [Tracepoint event]
```

**Comment by mbalabash** [comment]
It looks like this API depends on PMU/PMC metrics which not available in virtualized environments.

In fact, this not only depends on environment but on your CPU device too.

More context:
- [PMU/PMC](http://rts.lab.asu.edu/web_438/project_final/Talk%209%20Performance%20Monitoring%20Unit.pdf)
- [similar problem](https://www.researchgate.net/post/Why_doesnt_perf_report_cache-refernces_cache-misses).
- [linux perf basics](http://www.brendangregg.com/perf.html)

**Comment by mbalabash** [comment]
I think we need an opinion from someone who more deeply in these things. Maybe there is some workaround or I miss something.

**Comment by ai** [comment]
Does it work on CI with Linux environment? It could be just Docker virtualization problem.

**Comment by mbalabash** [comment]
Tomorrow I gonna prepare a few special tests for CircleCI / TravisCI and then will provide the results here.

**Comment by mbalabash** [comment]
@ai 

I made a few tests and run them on [CircleCI](https://app.circleci.com/jobs/github/mbalabash/estimo/98) and [TravisCI](https://travis-ci.org/mbalabash/estimo/jobs/645061835).

`rti-api › [rti-api]: should throw when RTI API does not support by the platform` - always pass

`rti-api › [rti-api]: should contain "tidelta" and "ticount" if RTI API supported by the platform` - always fails

So, it isn't working :(

**Comment by mitermayer** [comment]
@mbalabash @ai Is there any more information around this problem? 

Would love if we could get a solution for this issue

For future reference, if folks try to debug this issue, this is the thread_count feature commit on chromium https://github.com/webrtc-uwp/chromium-base/commit/ffcfa2d196748f0301c46430ac6ff05778d23244

**Comment by mitermayer** [comment]
This may help on investigation, when runing on bare metal Linux (The one that works and displays the tidelta correctly we have this execution header): 

```
{
  "chrome-bitness": 64,
  "clock-domain": "LINUX_CLOCK_MONOTONIC",
  "command_line": "/opt/google/chrome/google-chrome --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-features=TranslateUI --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --headless --hide-scrollbars --mute-audio --enable-thread-instruction-count --no-sandbox --disable-setuid-sandbox --disable-dev-shm-usage --remote-debugging-port=0 --user-data-dir=/tmp/puppeteer_dev_chrome_profile-TOVbOO --use-gl=swiftshader-webgl --allow-pre-commit-input about:blank",
  "cpu-brand": "Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz",
  "cpu-family": 6,
  "cpu-model": 158,
  "cpu-stepping": 13,
  "gpu-devid": 65535,
  "gpu-driver": "4.1.0.7",
  "gpu-gl-renderer": "Google SwiftShader",
  "gpu-gl-vendor": "Google Inc.",
  "gpu-psver": "3.00",
  "gpu-venid": 65535,
  "gpu-vsver": "3.00",
  "highres-ticks": 1,
  "network-type": "WiFi",
  "num-cpus": 12,
  "os-arch": "x86_64",
  "os-name": "Linux",
  "os-version": "5.5.13-arch2-1",
  "perfetto_trace_stats": {
    "buffer_stats": [
      {
        "buffer_size": 209715200,
        "bytes_overwritten": 0,
        "bytes_read": 458752,
        "bytes_written": 622592,
        "chunks_committed_out_of_order": 0,
        "chunks_discarded": 0,
        "chunks_overwritten": 0,
        "chunks_read": 14,
        "chunks_rewritten": 0,
        "chunks_written": 19,
        "padding_bytes_cleared": 0,
        "padding_bytes_written": 0,
        "patches_failed": 0,
        "patches_succeeded": 0,
        "readaheads_failed": 0,
        "readaheads_succeeded": 0,
        "trace_writer_packet_loss": 0,
        "write_wrap_count": 0
      }
    ],
    "chunks_discarded": 0,
    "data_sources_registered": 11,
    "data_sources_seen": 4,
    "patches_discarded": 0,
    "producers_connected": 5,
    "producers_seen": 0,
    "total_buffers": 1,
    "tracing_sessions": 1
  },
  "physical-memory": 31744,
  "product-version": "HeadlessChrome/80.0.3987.163",
  "trace-capture-datetime": "2020-5-22 2:22:7",
  "trace-config": "{\"enable_argument_filter\":false,\"enable_systrace\":false,\"excluded_categories\":[\"*\"],\"included_categories\":[\"devtools.timeline\",\"v8.execute\",\"toplevel\",\"blink.console\",\"blink.user_timing\",\"latencyInfo\",\"disabled-by-default-devtools.timeline\",\"disabled-by-default-devtools.timeline.frame\",\"disabled-by-default-devtools.timeline.stack\",\"disabled-by-default-v8.cpu_profiler\",\"disabled-by-default-v8.cpu_profiler.hires\"],\"record_mode\":\"record-until-full\"}",
  "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/80.0.3987.163 Safari/537.36",
  "v8-version": "8.0.426.30"
}
```
Running it from a docker container WITH PRIVILLEGES and without "tidelta" on the response we get this:

```
{
  "chrome-bitness": 64,
  "clock-domain": "LINUX_CLOCK_MONOTONIC",
  "command_line": "/usr/bin/google-chrome --disable-background-networking --enable-features=NetworkService,NetworkServiceInProcess --disable-background-timer-throttling --disable-backgrounding-occluded-windows --disable-breakpad --disable-client-side-phishing-detection --disable-component-extensions-with-background-pages --disable-default-apps --disable-dev-shm-usage --disable-features=TranslateUI --disable-hang-monitor --disable-ipc-flooding-protection --disable-popup-blocking --disable-prompt-on-repost --disable-renderer-backgrounding --disable-sync --force-color-profile=srgb --metrics-recording-only --no-first-run --enable-automation --password-store=basic --use-mock-keychain --headless --hide-scrollbars --mute-audio --enable-thread-instruction-count --no-sandbox --disable-setuid-sandbox --disable-dev-shm-usage --remote-debugging-port=0 --user-data-dir=/tmp/puppeteer_dev_chrome_profile-5xPBH6 --use-gl=swiftshader-webgl --allow-pre-commit-input about:blank",
  "cpu-brand": "Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz",
  "cpu-family": 6,
  "cpu-model": 158,
  "cpu-stepping": 13,
  "gpu-devid": 65535,
  "gpu-driver": "4.1.0.7",
  "gpu-gl-renderer": "Google SwiftShader",
  "gpu-gl-vendor": "Google Inc.",
  "gpu-psver": "3.00",
  "gpu-venid": 65535,
  "gpu-vsver": "3.00",
  "highres-ticks": 1,
  "network-type": "Ethernet",
  "num-cpus": 12,
  "os-arch": "x86_64",
  "os-name": "Linux",
  "os-version": "5.5.13-arch2-1",
  "physical-memory": 31744,
  "product-version": "HeadlessChrome/84.0.4147.13",
  "trace-capture-datetime": "2020-5-22 2:34:7",
  "trace-config": "{\"enable_argument_filter\":false,\"enable_systrace\":false,\"excluded_categories\":[\"*\"],\"included_categories\":[\"devtools.timeline\",\"v8.execute\",\"toplevel\",\"blink.console\",\"blink.user_timing\",\"latencyInfo\",\"disabled-by-default-devtools.timeline\",\"disabled-by-default-devtools.timeline.frame\",\"disabled-by-default-devtools.timeline.stack\",\"disabled-by-default-v8.cpu_profiler\",\"disabled-by-default-v8.cpu_profiler.hires\"],\"record_mode\":\"record-until-full\"}",
  "trace_processor_stats": {
    "android_log_num_failed": 0,
    "android_log_num_skipped": 0,
    "android_log_num_total": 0,
    "clock_sync_cache_miss": 31,
    "clock_sync_failure": 0,
    "compact_sched_has_parse_errors": 0,
    "compact_sched_switch_skipped": 0,
    "compact_sched_waking_skipped": 0,
    "counter_events_out_of_order": 0,
    "empty_chrome_metadata": 0,
    "ftrace_bundle_tokenizer_errors": 0,
    "fuchsia_invalid_event": 0,
    "fuchsia_non_numeric_counters": 0,
    "fuchsia_timestamp_overflow": 0,
    "gpu_counters_invalid_spec": 0,
    "gpu_counters_missing_spec": 0,
    "gpu_render_stage_parser_errors": 0,
    "graphics_frame_event_parser_errors": 0,
    "guess_trace_type_duration_ns": 1392,
    "heap_graph_location_parse_error": 0,
    "heap_graph_non_finalized_graph": 0,
    "heapprofd_missing_packet": 0,
    "heapprofd_non_finalized_profile": 0,
    "interned_data_tokenizer_errors": 0,
    "invalid_clock_snapshots": 0,
    "invalid_cpu_times": 0,
    "json_tokenizer_failure": 0,
    "meminfo_unknown_keys": 0,
    "metatrace_overruns": 0,
    "mismatched_sched_switch_tids": 0,
    "misplaced_end_event": 0,
    "mm_unknown_type": 0,
    "ninja_parse_errors": 0,
    "packages_list_has_parse_errors": 0,
    "packages_list_has_read_errors": 0,
    "parse_trace_duration_ns": 2222259,
    "perf_samples_skipped": 0,
    "perf_samples_skipped_dataloss": 0,
    "power_rail_unknown_index": 0,
    "proc_stat_unknown_counters": 0,
    "process_tracker_errors": 0,
    "rss_stat_negative_size": 0,
    "rss_stat_unknown_keys": 0,
    "rss_stat_unknown_thread_for_mm_id": 0,
    "sched_switch_out_of_order": 0,
    "sched_waking_out_of_order": 0,
    "slice_out_of_order": 0,
    "stackprofile_invalid_callstack_id": 0,
    "stackprofile_invalid_frame_id": 0,
    "stackprofile_invalid_mapping_id": 0,
    "stackprofile_invalid_string_id": 0,
    "stackprofile_parser_error": 0,
    "systrace_parse_failure": 0,
    "task_state_invalid": 0,
    "thread_time_in_state_out_of_order": 0,
    "thread_time_in_state_unknown_cpu_freq": 0,
    "tokenizer_skipped_packets": 0,
    "traced_buf": [
      {
        "buffer_size": 209715200,
        "bytes_overwritten": 0,
        "bytes_read": 557056,
        "bytes_written": 688128,
        "chunks_committed_out_of_order": 0,
        "chunks_discarded": 0,
        "chunks_overwritten": 0,
        "chunks_read": 17,
        "chunks_rewritten": 0,
        "chunks_written": 21,
        "padding_bytes_cleared": 0,
        "padding_bytes_written": 0,
        "patches_failed": 0,
        "patches_succeeded": 0,
        "readaheads_failed": 0,
        "readaheads_succeeded": 0,
        "trace_writer_packet_loss": 0,
        "write_wrap_count": 0
      }
    ],
    "traced_chunks_discarded": 0,
    "traced_data_sources_registered": 11,
    "traced_data_sources_seen": 4,
    "traced_patches_discarded": 0,
    "traced_producers_connected": 5,
    "traced_producers_seen": 0,
    "traced_total_buffers": 1,
    "traced_tracing_sessions": 1,
    "track_event_parser_errors": 0,
    "track_event_tokenizer_errors": 0,
    "vmstat_unknown_keys": 0,
    "vulkan_allocations_invalid_string_id": 0
  },
  "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/84.0.4147.13 Safari/537.36",
  "v8-version": "8.4.371.7"
}
```

Key interesting differences is that in the "bare metal" we get **perfetto_trace_stats** and on docker we get **trace_processor_stats**

**Comment by mitermayer** [comment]
It may seem like within docker we are getting into this path https://github.com/google/perfetto/blob/master/src/trace_processor/export_json.cc#L166

**Comment by mitermayer** [comment]
Looks like what controls if **trace_processor_stats** or **perfetto_trace_stats** is used was defined in here https://github.com/google/perfetto/commit/bb0e154023150a87068ea3c5172eb51afba95f0c

**Comment by mitermayer** [comment]
It seems that the **trace_processor_stats** could have been introduced in https://github.com/google/perfetto/commit/e2b4e5c6c0da4ca0d9d3aeaeb562a95c3b6b5844 But it's unclear to me still why within docker we get one over the other

**Comment by mitermayer** [comment]
I just noticed that I was comparing two different chrome versions:

```

  "os-version": "5.5.13-arch2-1",
  "product-version": "HeadlessChrome/80.0.3987.163",
  "v8-version": "8.0.426.30"

  "os-version": "5.5.13-arch2-1",
  "product-version": "HeadlessChrome/84.0.4147.13",
  "v8-version": "8.4.371.7"
```

Will update chrome and test again

**Comment by mbalabash** [comment]
Hey @mitermayer
Happy that you took attention to this issue. It would be awesome if we could achieve some progress here.

Since the last post I haven't had any updates about this issue.

My main hypothesis is that docker can't get data from `Performance Monitoring Counters`.

**Comment by mbalabash** [comment]
Hey @mittermayr!
Did you have any progress with your idea?

**Comment by MurzNN** [comment]
As described in article [JavaScript component-level CPU costs](https://calendar.perfplanet.com/2019/javascript-component-level-cpu-costs/) - this feature was work well on Chrome 78, so maybe try to retest this in exact 78 version of Chromium, and in later version this feature is changed/broken?

**Comment by mbalabash** [comment]
@MurzNN As we know from the article, that code was executed on the bare metal linux machine. And as I know, it wasn't tested in docker environment.

Also, all tests at the beginning of the story were with Chrome 78.

**Comment by MurzNN** [comment]
But `ticount` not shown even on my bare metal linux machine (Intel(R) Core(TM) i3-4330 CPU @ 3.50GHz, using Ubuntu Linux 16.04), so the source of this issue seems not in Docker, but in Chromium itself?

**Comment by mbalabash** [comment]
@MurzNN Which version of Chrome do you use? 

**Could you please run this example on your machine**:
```js
const puppeteer = require('puppeteer')

;(async () => {
  const browser = await puppeteer.launch({
    args: ['--no-sandbox', '--enable-thread-instruction-count'],
  })
  const page = await browser.newPage()

  await page.tracing.start({ path: './trace.json' })
  await page.goto('https://www.example.com/')
  await page.tracing.stop()

  await browser.close()
})()
```

**Then**:
1. Copy content from `trace.json` file
2. Create gist and paste content from the previous step
3. Provide a link to this gist in your comment.

**Comment by MurzNN** [comment]
Done using HeadlessChrome/83.0.4103.0 on native Ubuntu 20.04 Linux without docker: https://gist.github.com/MurzNN/5e2683d1c4e2e59d8478b4cc95c4366e

**Comment by mbalabash** [comment]
@MurzNN Yes, you right. Your trace file don't contain `ticount` and `tidelta` values.
It means that something has been changed in the latest Chrome versions. 

I not familiar with changes which were happened in `81, 82, 83` versions. So, I'm going to do homework and will provide an update ASAP.

**Comment by mbalabash** [comment]
> the source of this issue seems not in Docker, but in Chromium itself?

It's hard to be sure at 100%. But my main hypothesis is that a problem could be in hypervisors or hardware (hope I wrong).

Looks like a ton of changes were applied to [tracing feature](https://chromium-review.googlesource.com/q/hashtag:%22tracing%22+(status:open%20OR%20status:merged)). One of them could possibly break `enable-thread-instruction-count` API.

@MurzNN Could you please do [the test](https://github.com/puppeteer/puppeteer/issues/5223#issuecomment-576448290), which Andrey did on his bare metal Linux machine?
That test will run Chrome 79 and `enable-thread-instruction-count` API should work as expected.

**Comment by MurzNN** [comment]
@mbalabash, I do test on Ubuntu 20.04, output is here: https://gist.github.com/MurzNN/52c8944c747d026d8afee509537c7a75 - ticount is missing :( Same results is with Chromium 78 (r686622) - I change them in `cripts/chromeDetection.js`.

**Comment by MurzNN** [comment]
When I manually launch chrome and enable tracing, I see in output
```
$ ./chrome --no-sandbox --enable-thread-instruction-count
[329820:329820:0609/065603.823454:ERROR:sandbox_linux.cc(369)] InitializeSandbox() called with multiple threads in process gpu-process.
[329820:329820:0609/065605.305150:ERROR:buffer_manager.cc(488)] [.DisplayCompositor]GL ERROR :GL_INVALID_OPERATION : glBufferData: <- error from previous GL command
tracing_service_impl.cc: Configured tracing, #sources:2, duration:0 ms, #buffers:1, total buffer size:102400 KB, total sessions:1
[329975:329977:0609/065638.525405:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329975:329975:0609/065638.524778:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329822:329856:0609/065638.528999:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329975:329978:0609/065638.528997:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329822:329853:0609/065638.529197:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329820:329847:0609/065638.535137:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
[329894:329894:0609/065638.535123:ERROR:thread_instruction_count.cc(65)] perf_event_open failed, omitting instruction counters
```
Seems `perf_event_open failed` is source of problem?

And magic command "allow kernel profiling for non-privileged users", that I find [here](https://github.com/facebook/hhvm/issues/8026#issuecomment-365708696) - seems help:
```
echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid
```
After enabling it - I see `ticount` in json results of your test file!

**Comment by mbalabash** [comment]
@MurzNN I had the same behavior with `perf_event_open` on my Mac laptop. You can read all thread from [start](https://github.com/puppeteer/puppeteer/issues/5223#issuecomment-572565802). 

> After enabling it - I see ticount in json results of your test file!

Which case works for you? [First](https://github.com/puppeteer/puppeteer/issues/5223#issuecomment-637426299) or [Second](https://github.com/puppeteer/puppeteer/issues/5223#issuecomment-576448290)? Maybe both?

**Comment by MurzNN** [comment]
@mbalabash after `echo 1 | sudo tee /proc/sys/kernel/perf_event_paranoid` - both becomes work, and all other cases too!

**Comment by mbalabash** [comment]
@MurzNN Cool. It means that it still works on bare metal Linux )

And it probably should works inside a docker under bare metal Linux. If you allow docker to use `perf_event_open` ([possible solutions](https://github.com/puppeteer/puppeteer/issues/5223#issuecomment-573269857)).

Is it possible for you to test this case?

**Comment by MurzNN** [comment]
I test this in Docker - ticount shows well with `docker run --cap-add SYS_ADMIN myImage` (and even with `echo 3 | sudo tee /proc/sys/kernel/perf_event_paranoid`), without this argument - ticount missing. 

Also `ticount` not available in all variants on virtual environment, running same code natively or via Docker on several cloud VPS's, that I have, not show me `ticount`.

**Comment by MurzNN** [comment]
Here is my quick script for check `ticount` support:
```
const puppeteer = require('puppeteer');
const fs = require('fs');

puppeteerStart = async function() {
  const browser = await puppeteer.launch({
  args: [
    '--no-sandbox',
    '--enable-thread-instruction-count',
  ]});
  const page = await browser.newPage();
  await page.tracing.start({path: 'trace.json'});
  await page.goto('https://google.com/');
  await page.tracing.stop();
  await browser.close();
  let traceData = fs.readFileSync('trace.json').toString();
  // console.log(traceData);
  if(traceData.includes(',"ticount":')) {
    console.log('ticount support is enabled!');
  } else {
    console.log('ticount support is disabled!');
  }
}

puppeteerStart();
```

**Comment by mbalabash** [comment]
@MurzNN Thank you for your work. It is really helpful to understand this issue.

Your tests also confirm the hypothesis that the problem is in hypervisors. 
Which means we can't use it in CI 😔.

**Comment by MurzNN** [comment]
So, is there any improvements about ticount support checking planned for puppeteer?

**Comment by MurzNN** [comment]
It's confirmed, just waiting for the response!

**Comment by OrKoN** [comment]
>  So, is there any improvements about ticount support checking planned for puppeteer?

I don't think we plan anything in this area and it looks like the issue for the original reporter is resolved (skimming over the comments). Also, Puppeteer just forwards the traces from the browser so if there is a problem with the tracing data it's better to open an issue at crbug.com.

**Comment by OrKoN** [comment]
So I am going to mark this one as not planned, because we are unlikely to do anything about it in Puppeteer. I recommend raising the issue at crbug.com with the owners of the --enable-thread-instruction-count flag in Chromium.

**Comment by sophiebits** [comment]
For anyone else curious – it appears that support for these fields was removed in Chrome in this Jun 2022 commit, which landed in release 105.0.5137.0:

https://chromium.googlesource.com/chromium/src/+/736be85b0d7c4fbfaacc7fcd8571828f8ad4dee1
